src/main/typescript/wcardinal/ui/d-tree.ts

```diff
@@ -3,19 +3,429 @@
+						}
+					}
+
+					const isParent: boolean = item.children && ( item.children.length > 0 );
+					const text = item.text ? item.text : "";
+					if( expandAll != null ) {
+						item.expanded = expandAll;
+					} else if ( item.expanded == null ) {
+						item.expanded = false; // set default expand status of item is false
+					}
+					const itemImage = item.image ? item.image : null;
+					let isItemExisted = false;
+					let itemOptions = this._itemOptions.get( item );
+
+					if( itemOptions != null ) {
+						itemOptions.rawData = item;
+						itemOptions.text = text;
+						itemOptions.y = this._itemY;
+						itemOptions.isParent = isParent;
+						itemOptions.expanded = item.expanded;
+						itemOptions.image = itemImage;
+						isItemExisted = true;
+					} else {
+						itemOptions = {
+							rawData: item,
+							text,
+							level,
+							y: this._itemY,
+							isParent,
+							expanded: item.expanded,
+							image: itemImage
+						};
+					}
+
+					/* displayed items need to satisfy 1 of the 2 conditions:
+					1. is root item
+					2. the parent item is show and expand
+					*/
+					if ( parentItemOptions == null ||
+						( parentItemOptions &&
+							parentItemOptions.expanded &&
+							parentItemOptions.showable ) ) {
+						itemOptions.showable = true;
+						this._itemOptionsShowable.push( itemOptions );
+						this._itemY += this._itemHeight;
+					} else {
+						itemOptions.showable = false;
+					}
+					if ( !isItemExisted ) {
+						this._itemOptions.set( item, itemOptions );
+					}
+					if ( item && item.children ) {
+						this.updateData( itemOptions, item, level + 1, expandAll );
+					}
+				}
+			}
+		}
+
+		onKeyDown( e: KeyboardEvent ): boolean {
+			this._keyboardEvent = e;
+			return super.onKeyDown( e );
+		}
+
+		onKeyUp( e: KeyboardEvent ): boolean {
+			this._keyboardEvent = e;
+			return super.onKeyUp( e );
+		}
+
+		protected onSelect( selection: DTreeItemRawData ) {
+			// multi select by "ctr" key + click
+			if( this._keyboardEvent && this._keyboardEvent.ctrlKey ) {
+				this._selection.toggle( selection );
+
+				// multi select by "shift" key + click
+			} else if ( this._keyboardEvent && this._keyboardEvent.shiftKey ) {
+				const lastSelection = this._selection.get( this._selection.size() - 1 );
+				if( lastSelection ) {
+					this._selection.clear();
+					const selectionY = Number( this._itemOptions.get( selection )?.y );
+					const lastSelectionY = Number( this._itemOptions.get ( lastSelection )?.y );
+					const maxY = selectionY < lastSelectionY ?
+						lastSelectionY - this._itemHeight :
+						selectionY;
+					const minY = selectionY < lastSelectionY ?
+						selectionY :
+						lastSelectionY + this._itemHeight;
+
+					this._itemOptionsShowable.every( ( item ) => {
+						if( item.y >= minY && item.y <= maxY && item.showable) {
+							this._selection.add( item.rawData );
+						}
+						return item.y < maxY;
+					} );
+					this._selection.add( lastSelection );
```

Comment:
`lastSelection` does not need to be added.

Reply:
I use lastSelection as an anchor when selecting multiple items with "shift" + click.
For example, I have a tree widget with items 1 to 10.
I select item 5 and then I "shift" + click on item 7, then item 5 6 7 will be selected.
After that, I "shift", click on item 3, then item 3 4 5 will be selected.

Reply:
Let me make sure the shift+click selection behavior. When No.5 is selected, if you click No.7 with the shift modifier, No. 5, 6 and 7 will be selected. And then if you click the No. 3 with the shift modifier, No. 3, 4 will be selected additionally. Namely, No. 3,4,5,6 and 7 will be selected. Am I right?

Reply:
> And then if you click the No. 3 with the shift modifier, No. 3, 4 will be selected additionally. Namely, No. 3,4,5,6 and 7 will be selected. Am I right?

No. 3, 4, 5 will be selected, and No.6, 7 will be un-select.
I refer the shift+click selection behavior of Win 10 File Explorer.

Reply:
Good to know. Thank you.